_posts/2024-08-06-quarkus-and-leyden.adoc

```diff
@@ -0,0 +1,303 @@
+
+.When doing training runs, we create an archive that contains information on how the application runs. This archive includes not only classes from the core libraries, but also classes from our application.
+image::AoT_vs_JiT_dynamic.svg[Dynamic CDS benefits,float="right",align="center"]
+
+As with lazy loading, this lazy approach results in almost the exact same linkage being established on every run. The time spent stopping and restarting execution to lazily connect the class graph comprises a noticeable percentage of JDK startup, application startup and application warm up (time to peak running). We could speed up startup and, more crucially, warm up time if we could pre-compute this linkage and avoid the need to establish it at runtime.
+
+=== Synergy with Quarkus
+
+Loading and linking of classes is an important step in the warm up of the application because it involves searching through the whole classpath for all classes and objects referenced by the bytecode the JVM is going to run. By default, this is done as a lazy operation because loading and linking all existing classes in the classpath would not only require a bigger memory footprint, but also a bigger warm up time. This is why the JVM only compiles and links the bytecode that is going to be used.
+
+This is a process that Quarkus already speeds up by, among other strategies, aggressively reducing the set of classes included in the classpath, so the search for matches is faster. The search for classes is also accelerated by indexes which Quarkus can generate when it fully analyzes the application at build time. But it is still a heavy operation that is difficult to execute ahead of time, before we know what is going to be run and how. Quarkus might be able to provide some additional hints to the linker in the future.
+
+The first improvement Leyden is offering to improve startup time is to upgrade the AOT model originally develped as part of the CDS project to encompass not just pre-loading of classes but also pre-linking, as described in https://openjdk.org/jeps/8315737[JEP Ahead-of-Time Class Linking].
+
+An AOT Cache can be generated during a training run that bootstraps the JVM and, optionally, executes application-specific code.
+As with a CDS archive, the AOT Cache stores a class graph for all classes loaded during the training run in a format that allows it to be quickly remapped on a subsequent run. The stored graph also includes any linkage information established by code executed during the training run. Pre-cached links avoid the need to stop and start execution to perform linkage on subsequent runs.
+
```

Comment:
I think we should reference the AOT Archive diagram from here rather than later. It is a little early in the sense that we have not yet talked about AOT vs JIT compilation. However, we have now introduced AOT vs JIT linking and the picture helps to clarify what has been said so far. The fact that compilation is also included, as indicated by the diagram, is not going to be very surprising to those who notice it.

Reply:
I've moved the image to just after "The Leyden project has already sucessfully prototyped shifting the work of method compilation from JIT to AOT. Execution and compilation of methods is tracked during the training run. At the end of the run any associated profiling information and compiled code for the method are saved to the AOT Cache, allowing them to be quickly mapped back into memory and reused when the application is next run."

Good compromise? And breaks the flow better.

Reply:
I think that is still too late. I would prefer to place it just before the heading "JIT vs AOT Compilation". There is a lot of text in that section about AOT vs JIT compilation which would need to be worked through before reaching the location where you have put it. Placing the picture before getting into any talk about compilation clarifies in an image what has just been explained in text about AOT vs JIT linkage. It also presents this new step of AOT vs JIT linkage pictorially in the same terms as what was said pictorially regarding AOT vs JIT loading. So, it reinforces the idea that this is more of the same in language and image. I think that helps swallow this step and also makes it easier to swallow the next step where we explain AOT vs JIT compilation as once again being more of the same.

Reply:
Moved to where @adinn wanted it.